[dynamo] Automatically convert loop bodies to function calls

Fidget-Spinner opened the pull request:
Related to https://github.com/pytorch/pytorch/issues/111441, but will not fix it until we implement function caching/higher order operators for for loops.

**See below for results, roughly 10% compilation speedup for medium loops, 20% compilation speedup for large loops.**

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Fidget-Spinner committed:
Convert loop bodies to function calls

Fidget-Spinner committed:
fix some stuff

Fidget-Spinner commented:
@jansel or @albanD do you think this is the right path for utilizing function caching to speed up compile times? Runtime performance should be identical.

jansel reviewed (changes requested):
1) I worry this AST rewriting approach won't correctly handle mutating local variables and closures.
2) `NestedUserFunctionVariable(...).call_function(...)` likely won't save on compile times, since we will stil retrace the loop body `N` times.  In order to save on compile times we will need to introduce a graph break or higher order op.
3) I don't think we should do this all of the time, we should only do it for very large loops.

Fidget-Spinner commented:
Thank you for the review.

> 1. I worry this AST rewriting approach won't correctly handle mutating local variables and closures.

I worry about closures too. I will run some tests.

> 2. `NestedUserFunctionVariable(...).call_function(...)` likely won't save on compile times, since we will stil retrace the loop body `N` times.  In order to save on compile times we will need to introduce a graph break or higher order op.

Yeah. This PR won't speed up anything unless we either start caching function compiles, or introduce a simple higher order op. I will look into either one.

> 3. I don't think we should do this all of the time, we should only do it for very large loops.

Agreed.

Fidget-Spinner commented:
I realised the higher order operations and subgraph tracing mechanism can already support this. So I will use that instead of AST rewriting.

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner committed:
use bytecode transformations instead of AST

Fidget-Spinner commented:
Results:

For the following code:
```python
import torch

# @torch.compile
# def foo(x):
#     for i in range(3):
#         x = x + 1
#     return x

@torch.compile
def foo(x):
    for i in range(2000):
        x = x + 1
    return x

foo(torch.randn(1))
```

Results:
```
main@afadfa0175221b0bb57af6f16fa2bb8210be7902:
real    0m8.595s
user    0m8.036s
sys     0m0.631s

this branch:
real    0m1.967s
user    0m1.849s
sys     0m0.191s
```
Roughly a 4x speedup. Now I need to look at whether the generated code is any worse.

Fidget-Spinner commented:
For the following code:
```
import torch

@torch.compile
def foo(x):
    for i in range(2000):
        x = x + 1
    return x

y = torch.randn(1)
print(y)
print(foo(y))

```
Results:
```
main@afadfa0175221b0bb57af6f16fa2bb8210be7902
tensor([-0.3247])
tensor([1999.6753])

real    0m9.733s
user    0m9.127s
sys     0m0.677s

my branch@HEAD
tensor([1.0786])
tensor([2001.0786])

real    0m8.269s
user    0m7.604s
sys     0m0.737s


Increasing the loop iterations to 6000:

main@afadfa0175221b0bb57af6f16fa2bb8210be7902
tensor([0.0062])
tensor([6000.0063])

real    0m26.017s
user    0m24.887s
sys     0m1.203s

my branch@HEAD
tensor([0.4038])
tensor([6000.4038])

real    0m20.035s
user    0m18.729s
sys     0m1.348s
```

Fidget-Spinner committed:
make it actually return results

Fidget-Spinner commented:
Note: loop compilation is currently slightly worse than linear in dynamo+inductor. It should become sub-linear with this PR for simple loops, since this PR bypasses quite a bit of stuff. Once caching or a higher order op is implemented, it should become constant.

The main reason it's slightly worse than linear is because of this https://github.com/pytorch/pytorch/issues/111686#issuecomment-1774123924. If I understand correctly.

It has been some time since I looked at the originating course, so I may be wrong.

Fidget-Spinner committed:
docs, only try converting once

Fidget-Spinner committed:
lint

Fidget-Spinner committed:
fix tests, and lint

Fidget-Spinner committed:
fix tests

Fidget-Spinner committed:
fix none access

Fidget-Spinner commented:
With num_steps=150 for the program in the user's bug report here at https://github.com/pytorch/pytorch/issues/111441:

```
main@afadfa0175221b0bb57af6f16fa2bb8210be7902
17787.90   compiled

my branch@HEAD
16076.84   compiled
```
Roughly 10% faster.

Fidget-Spinner committed:
set the right flag

Fidget-Spinner committed:
block exceptions as well

Fidget-Spinner committed:
lint, support 3.8 - 3.10

voznesenskym commented:
Very cool, can't believe I missed this PR. I was just playing around with something like this, but far less along. LMK if you need a hand or want to talk through any of it.

voznesenskym commented:
I know this passes tests, but it would be nice to see torturous examples added to tests on purpose: Loops of loops of loops, loops of inling, etc.

voznesenskym commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1152,6 +1169,279 @@ def call_function(
+    ):
+        from ..source import GlobalSource
+
+        super().__init__(value, source, **kwargs)
+        self.store_target = -1
+        self.source = GlobalSource("range")
```
Nit - this will break in stupid ways if there is shadowing, consider firmly using builtins.range

voznesenskym replied:
It's weird to take an optional source arg, pass it to super, and then blow out the self.source like this. I would just reject and assert that no source is ever passed here.

voznesenskym commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1152,6 +1169,279 @@ def call_function(
+    ):
+        from . import ConstantVariable
+
+        if len(value.unpack_var_sequence(tx)) < 100:
+            raise CannotConvertRangeToHigherOrder(
+                "Loop too small to consider optimizing"
+            )
```
Nit - contrived heuristic! Move it to config or lift to a constant and document it at least...

voznesenskym commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1152,6 +1169,279 @@ def call_function(
+            raise CannotConvertRangeToHigherOrder(
+                "Control flow too complex for loop higher order op."
+            )
+        assert loop_body_instructions[0].opname == "STORE_FAST"
```
Consider a comment explaining why after for_iter, it's always STORE_FAST

voznesenskym commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1152,6 +1169,279 @@ def call_function(
+        assert loop_body_instructions[0].opname == "STORE_FAST"
+        co_code: List[int] = []
+        # We skip the first instruction as it's a STORE_FAST, while we skip the last as it's the JUMP_BACKWARD.
+        for inst in loop_body_instructions[1:-1]:
```
consider detecting empty loop body?

voznesenskym commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1152,6 +1169,279 @@ def call_function(
+                (body_r, body_spec),
+                body_graph,
+                body_lifted_freevars,
+            ) = speculate_subgraph(
```
Nice use of speculate_subgraph - the limitations this come with offer nice safety. Consider a comment here beyond just the graph break raise below.

voznesenskym commented:
Did a high level pass, nothing jumps out as bad, but not stamping, waiting for tests that at least exercise the >100 loop heuristic.

Fidget-Spinner committed:
Address review (TODO: specific tests)

Fidget-Spinner commented:
Thanks, I'm facing a problem right now with `speculate_subgraph`:
For the this type of code:

```
t = torch.tensor(1)
for i in range(x):
    t += i
```

Due to the `i` being passed to `speculate_subgraph` as a `ConstantVariable`, I suspect it just constant propagates the value in and doesn't use the value from the function arguments.

So on the next iteration when I adjust the `i`, it just reuses the same value (in the case 0 or 1, whatever the loop starts from, in the function body.

My immediate thought would be to convert `i` to a 0d `TensorVariable` and pass it in. But then the problem is what is the `Source` of something like that? Basically, I want to propagate and substitute i out as `torch.tensor(i)`.

I would be happy to get on a call to walk through the PR with you and seek your help on this part, since I'm admittedly still not the best at the whole `VariableTracker` stuff. I'll contact you on chat.

voznesenskym commented:
> Thanks, I'm facing a problem right now with `speculate_subgraph`:
> 
> For the this type of code:
> 
> 
> 
> ```
> 
> t = torch.tensor(1)
> 
> for i in range(x):
> 
>     t += i
> 
> ```
> 
> 
> 
> Due to the `i` being passed to `speculate_subgraph` as a `ConstantVariable`, I suspect it just constant propagates the value in and doesn't use the value from the function arguments.
> 
> 
> 
> So on the next iteration when I adjust the `i`, it just reuses the same value (in the case 0 or 1, whatever the loop starts from, in the function body.
> 
> 
> 
> My immediate thought would be to convert `i` to a 0d `TensorVariable` and pass it in. But then the problem is what is the `Source` of something like that? Basically, I want to propagate and substitute i out as `torch.tensor(i)`.
> 
> 
> 
> I would be happy to get on a call to walk through the PR with you and seek your help on this part, since I'm admittedly still not the best at the whole `VariableTracker` stuff. I'll contact you on chat.
> 
> 

I think this is a good example of some limitations of how we handle arg lifting. I don't know if making it a tensor off the get go is the right direction. I'll be home in an hour or so, I'll patch it down and play with it.

This potentially touches on another question - which is specialization of control flow hidden within ops. There might be inputs that change the graph that gets captured even with your banned op codes. We may need to visit how we handle guarding and code reuse here.

Fidget-Spinner committed:
fix docs, improve code a little, add failing tests

voznesenskym commented:
Okay, played with this for ~15 mins - the surface level problem is apparent. Assuming even 1 step ranges (not a good assumption, but we can fix that later), we are not capturing the range semantics anywhere. Your graph is snapshotting range step 0 (which, in the simple example, is 0).

 I propose that we:
 
 1) Make the range index position an argument to the range HOP
 2) Make the HOP body return an updated range value - which means accurately reflecting the range update in the graph body of the function representing a single iter step.
 
 Then, we need to go back and support ranges that step by k, ranges that range high->low, etc.

Fidget-Spinner commented:
Thanks for all your reviews so far!

> 1. Make the range index position an argument to the range HOP

It's not so obvious, but I already do that here https://github.com/pytorch/pytorch/pull/113538/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1393

> 2. Make the HOP body return an updated range value - which means accurately reflecting the range update in the graph body of the function representing a single iter step.

Again, pretty subtle, but here https://github.com/pytorch/pytorch/pull/113538/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1442

I have a hunch that because it sees `ConstantVariable`, it's just burning in the constant as a value or something. The original code of `speculate_subgraph` actually doesn't support `ConstantVariable` and strictly supports `TensorVariable` only.

Fidget-Spinner commented:
>  Assuming even 1 step ranges (not a good assumption, but we can fix that later), 

Oh good catch, this is indeed a bug that I missed. Should be easy to fix :).

voznesenskym commented:
> Thanks for all your reviews so far!
> 
> > 1. Make the range index position an argument to the range HOP
> 
> It's not so obvious, but I already do that here https://github.com/pytorch/pytorch/pull/113538/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1393
> 
> > 2. Make the HOP body return an updated range value - which means accurately reflecting the range update in the graph body of the function representing a single iter step.
> 
> Again, pretty subtle, but here https://github.com/pytorch/pytorch/pull/113538/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1442
> 
> I have a hunch that because it sees `ConstantVariable`, it's just burning in the constant as a value or something. The original code of `speculate_subgraph` actually doesn't support `ConstantVariable` and strictly supports `TensorVariable` only.

I saw this - but I don't think that's being reflected correctly - I meant you need to manually lift it into an arg - you have to force it to behave like this, even if it's a constant.

voznesenskym commented:
> > Thanks for all your reviews so far!
> > > 1. Make the range index position an argument to the range HOP
> > 
> > 
> > It's not so obvious, but I already do that here https://github.com/pytorch/pytorch/pull/113538/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1393
> > > 2. Make the HOP body return an updated range value - which means accurately reflecting the range update in the graph body of the function representing a single iter step.
> > 
> > 
> > Again, pretty subtle, but here https://github.com/pytorch/pytorch/pull/113538/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1442
> > I have a hunch that because it sees `ConstantVariable`, it's just burning in the constant as a value or something. The original code of `speculate_subgraph` actually doesn't support `ConstantVariable` and strictly supports `TensorVariable` only.
> 
> I saw this - but I don't think that's being reflected correctly - I meant you need to manually lift it into an arg - you have to force it to behave like this, even if it's a constant.

To elaborate - we have:

```
for_loop_body_0 = self.for_loop_body_0
for_loop_wrapper = torch__dynamo_variables_higher_order_ops_for_loop_wrapper(for_loop_body_0, res, res, 0);
...
for_loop_wrapper_1 = torch__dynamo_variables_higher_order_ops_for_loop_wrapper(for_loop_body_0, getitem, getitem_1, 1)
```

We just need to do a little glue in the body ;) 

I have a PR that makes this pass, give me a moment, I can push to this one if you don't mind?

voznesenskym committed:
Properly create proxies for subgraph tracing consts

Fidget-Spinner commented:
> I have a PR that makes this pass, give me a moment, I can push to this one if you don't mind?

Sure go ahead. That would be greatly appreciated!

voznesenskym commented:
> > > Thanks for all your reviews so far!
> > > > 1. Make the range index position an argument to the range HOP
> > > 
> > > 
> > > It's not so obvious, but I already do that here https://github.com/pytorch/pytorch/pull/113538/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1393
> > > > 2. Make the HOP body return an updated range value - which means accurately reflecting the range update in the graph body of the function representing a single iter step.
> > > 
> > > 
> > > Again, pretty subtle, but here https://github.com/pytorch/pytorch/pull/113538/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1442
> > > I have a hunch that because it sees `ConstantVariable`, it's just burning in the constant as a value or something. The original code of `speculate_subgraph` actually doesn't support `ConstantVariable` and strictly supports `TensorVariable` only.
> > 
> > 
> > I saw this - but I don't think that's being reflected correctly - I meant you need to manually lift it into an arg - you have to force it to behave like this, even if it's a constant.
> 
> To elaborate - we have:
> 
> ```
> for_loop_body_0 = self.for_loop_body_0
> for_loop_wrapper = torch__dynamo_variables_higher_order_ops_for_loop_wrapper(for_loop_body_0, res, res, 0);
> ...
> for_loop_wrapper_1 = torch__dynamo_variables_higher_order_ops_for_loop_wrapper(for_loop_body_0, getitem, getitem_1, 1)
> ```
> 
> We just need to do a little glue in the body ;)
> 
> I have a PR that makes this pass, give me a moment, I can push to this one if you don't mind?

I don't want to push up to this without your go-ahead, so I made https://github.com/pytorch/pytorch/pull/117450

Feel free to collapse that into this - it fixes the first test to properly track what we care about, and actually makes it pass. 

Note - the approach of creating symints for constants to track them is meant to demonstrate how to use a VT with a proxy-able definition (as_proxy conforming) - alternatively, instead of a blanket conversion, we can actually do this outside of speculate_subgraph and pass-it-through-the-top. This is more conservative, and probably a better direction.

Fidget-Spinner commented:
> I don't want to push up to this without your go-ahead, so I made #117450

Woops, our messages crossed. I sent the go-ahead just before you sent that.

voznesenskym commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -152,8 +161,14 @@ def validate_args_and_maybe_create_graph_inputs(
-            tracer.create_graph_input("const")
-            new_arg = a
+            const_proxy = tracer.create_graph_input("const")
+            if isinstance(a, ConstantVariable) and proxy_for_constants:
+                sym_a = torch.SymInt(a.as_python_constant())
+                const_proxy.node.meta["example_value"] = sym_a
+                sym_arg = SymNodeVariable(const_proxy, sym_a)
+                new_arg = sym_arg
```
We should bikeshed on this. I don't have a favorite plan, so I am going to rant off a list of various ways to do this, with pros and cons.

Option 1: This. The benefit is that it opens the door for symbolic args nicely, and we have access to the lowest tracer needed here (current subtracer, aka the tracer input). The con is that its kinda hand-waivey, turning all constants into dynamic values may not be desirable (should always be sound tho). It may even increase the graph break space, meaning less things get captured and more things are likely to fail speculation for no good reason!  I don't really like it.

Option 2: making the sym outside of speculate - where we set store_target for iteration. The benefit is that we are only making a symbolic for what we need, which is nice. The con is that it is *very* annoying to get the subtracer early, and so, it's kind of a PITA to call create_graph_input earlier! We can call it later, but SymNodeVariable should be created with a proxy, and it makes an annoying chicken and egg problem.

Option 3: User defined mapping of which constants to turn into symints. No real downside, but kind of an annoying API to use, I guess?

Option 4: RangeIteratorVariable always holds a symbolic value, with a proxy that is a node outside of the speculated subgraph.

Option 5: Alt alt approach is we don't use symbolic shapes, but produce a new light-weight proxyable int, where we can shuffle the proxy around and have it balance being an input to the subgraph without being an input to the parent (must be a named node in the parent tho, a la maybe_fx_node() below)

Thoughts @rzou?

voznesenskym commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3729,6 +3732,106 @@ def test_non_aliasing_util(self):
             _assert_tensors_nonaliasing(a, a)
 
+    @torch._dynamo.config.patch(
+        for_loop_medium_size_boundary=10
```
@Fidget-Spinner I rewrote this to be something we can scope to the test, to make for more manageable/easier testing. Lmk if you hate it.

voznesenskym commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3729,6 +3732,106 @@ def test_non_aliasing_util(self):
+        self.assertEqual(cnt.frame_count, 1)
+        self.assertEqual(cnt.op_count, 29)
+
+    def test_loop_to_higher_order_namespace_consistent(self):
```
TODO: @voznesenskym or @Fidget-Spinner - rewrite this test like the one above?

Fidget-Spinner replied:
Done!

voznesenskym commented:
Updated, releasing lock, feel free to take it from here, I think I did the useful bit of unblocking you.

Fidget-Spinner committed:
add tests for step

Fidget-Spinner commented:
An interesting bug:
The following code
```
import time
import torch

import numpy as np

class timeit:
    def __init__(self, tag=""):
        self.tag = tag

    def __enter__(self):
        if torch.cuda.is_available():
            torch.cuda.synchronize()
        self.start = time.perf_counter()
        return self

    def __exit__(self, *args):
        if torch.cuda.is_available():
            torch.cuda.synchronize()
        self.end = time.perf_counter()
        interval_ms = 1000 * (self.end - self.start)
        print(f"{interval_ms:8.2f}   {self.tag}")

@torch.compile
def getNormsSq0(errors):
    return (errors*errors).sum(axis=1)


@torch.compile
def trajH0(h, B, alpha, num_steps):
    """(normalize for total(h)=1)
    
    Simulate trajectories of fixed point equation
     e=e-alpha x <e, x>

    where x is a Gaussian RV with diagonal covariance entries h

    p: power-law decay constant of covariance eigenvalues
    d: number of dimensions
    B: batch size (number of trajectories to sample)
    alpha: step size
    num_steps: how many steps to run for

    Returns (num_steps, B) vector of ||e||^2
    """

    # h = h / np.sum(h)
    hsqrt = np.sqrt(h)
    E = np.ones((B, d))
    E = np.random.randn(B, d)
    traj = np.zeros((num_steps, B, d))

    for step_idx in range(num_steps):
        X = hsqrt * np.random.randn(B, d)
        losses = np.einsum("BD,BD->B", E, X)
        E -= alpha * np.einsum("BD,B->BD", X, losses)
        traj[step_idx] = E

    return np.sum(traj * traj, axis=2)

d = 1000
h = np.arange(1, d + 1)
h = h ** -1.1

num_steps=150
d = h.shape[0]

a = 2/(2*h.max() + h.sum())

#with timeit("regular"):
#        means = getNormsSq0(trajH(h, B=1000, alpha=a, num_steps=num_steps))

with timeit("compiled"):
    means = getNormsSq0(trajH0(h, B=1000, alpha=a, num_steps=150))
```

Errors now with:
```
 File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 748, in step
    getattr(self, inst.opname)(inst)
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 1140, in FOR_ITER
    new_locals = op.functionalize(self)
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/higher_order_ops.py", line 1413, in functionalize
    ) = speculate_subgraph(
        ^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/higher_order_ops.py", line 271, in speculate_subgraph
    output = f.call_function(tx, args, sub_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/functions.py", line 252, in call_function
    return super().call_function(tx, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/functions.py", line 81, in call_function
    return tx.inline_user_function_return(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 655, in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 2286, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 2401, in inline_call_
    tracer.run()
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 785, in run
    and self.step()
        ^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 748, in step
    getattr(self, inst.opname)(inst)
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 467, in wrapper
    return inner_fn(self, inst)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 1825, in CALL
    self.call_function(fn, args, kwargs)
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 649, in call_function
    self.push(fn.call_function(self, args, kwargs))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/misc.py", line 878, in call_function
    return NumpyNdarrayVariable.create(tx, proxy)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/tensor.py", line 857, in create
    return wrap_fx_proxy_cls(
           ^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/builder.py", line 1387, in wrap_fx_proxy_cls
    example_value = get_fake_value(proxy.node, tx, allow_non_graph_fake=True)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1585, in get_fake_value
    raise TorchRuntimeError(str(e)).with_traceback(e.__traceback__) from None
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1546, in get_fake_value
    ret_val = wrap_fake_exception(
              ^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1087, in wrap_fake_exception
    return fn()
           ^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1547, in <lambda>
    lambda: run_node(tx.output, node, args, kwargs, nnmodule)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1652, in run_node
    raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1631, in run_node
    return node.target(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1943, in __call__
    out = self.f(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_numpy/random.py", line 46, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/ken/Documents/GitHub/pytorch/torch/_numpy/random.py", line 111, in randn
    values = torch.randn(size, dtype=dtype)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.TorchRuntimeError: Failed running call_function <Wrapped function <original randn>>(*(1000, 1000), **{}):
randn(): argument 'size' failed to unpack the object at pos 1 with error "AttributeError: 'int' object has no attribute 'is_int'"
```

It's traced back to the lines
```
            if isinstance(a, ConstantVariable) and proxy_for_constants and a.as_python_constant() is not None:
                sym_a = torch.SymInt(a.as_python_constant())
                const_proxy.node.meta["example_value"] = sym_a
                sym_arg = SymNodeVariable(const_proxy, sym_a)
                new_arg = sym_arg
```

Seems that the numpy compatibility layer is treating a `SymInt` as a `SymNode`? Not sure why here.

voznesenskym commented:
> An interesting bug:
> 
> The following code
> 
> ```
> 
> import time
> 
> import torch
> 
> 
> 
> import numpy as np
> 
> 
> 
> class timeit:
> 
>     def __init__(self, tag=""):
> 
>         self.tag = tag
> 
> 
> 
>     def __enter__(self):
> 
>         if torch.cuda.is_available():
> 
>             torch.cuda.synchronize()
> 
>         self.start = time.perf_counter()
> 
>         return self
> 
> 
> 
>     def __exit__(self, *args):
> 
>         if torch.cuda.is_available():
> 
>             torch.cuda.synchronize()
> 
>         self.end = time.perf_counter()
> 
>         interval_ms = 1000 * (self.end - self.start)
> 
>         print(f"{interval_ms:8.2f}   {self.tag}")
> 
> 
> 
> @torch.compile
> 
> def getNormsSq0(errors):
> 
>     return (errors*errors).sum(axis=1)
> 
> 
> 
> 
> 
> @torch.compile
> 
> def trajH0(h, B, alpha, num_steps):
> 
>     """(normalize for total(h)=1)
> 
>     
> 
>     Simulate trajectories of fixed point equation
> 
>      e=e-alpha x <e, x>
> 
> 
> 
>     where x is a Gaussian RV with diagonal covariance entries h
> 
> 
> 
>     p: power-law decay constant of covariance eigenvalues
> 
>     d: number of dimensions
> 
>     B: batch size (number of trajectories to sample)
> 
>     alpha: step size
> 
>     num_steps: how many steps to run for
> 
> 
> 
>     Returns (num_steps, B) vector of ||e||^2
> 
>     """
> 
> 
> 
>     # h = h / np.sum(h)
> 
>     hsqrt = np.sqrt(h)
> 
>     E = np.ones((B, d))
> 
>     E = np.random.randn(B, d)
> 
>     traj = np.zeros((num_steps, B, d))
> 
> 
> 
>     for step_idx in range(num_steps):
> 
>         X = hsqrt * np.random.randn(B, d)
> 
>         losses = np.einsum("BD,BD->B", E, X)
> 
>         E -= alpha * np.einsum("BD,B->BD", X, losses)
> 
>         traj[step_idx] = E
> 
> 
> 
>     return np.sum(traj * traj, axis=2)
> 
> 
> 
> d = 1000
> 
> h = np.arange(1, d + 1)
> 
> h = h ** -1.1
> 
> 
> 
> num_steps=150
> 
> d = h.shape[0]
> 
> 
> 
> a = 2/(2*h.max() + h.sum())
> 
> 
> 
> #with timeit("regular"):
> 
> #        means = getNormsSq0(trajH(h, B=1000, alpha=a, num_steps=num_steps))
> 
> 
> 
> with timeit("compiled"):
> 
>     means = getNormsSq0(trajH0(h, B=1000, alpha=a, num_steps=150))
> 
> ```
> 
> 
> 
> Errors now with:
> 
> ```
> 
>  File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 748, in step
> 
>     getattr(self, inst.opname)(inst)
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 1140, in FOR_ITER
> 
>     new_locals = op.functionalize(self)
> 
>                  ^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/higher_order_ops.py", line 1413, in functionalize
> 
>     ) = speculate_subgraph(
> 
>         ^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/higher_order_ops.py", line 271, in speculate_subgraph
> 
>     output = f.call_function(tx, args, sub_kwargs)
> 
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/functions.py", line 252, in call_function
> 
>     return super().call_function(tx, args, kwargs)
> 
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/functions.py", line 81, in call_function
> 
>     return tx.inline_user_function_return(
> 
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 655, in inline_user_function_return
> 
>     return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
> 
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 2286, in inline_call
> 
>     return cls.inline_call_(parent, func, args, kwargs)
> 
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 2401, in inline_call_
> 
>     tracer.run()
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 785, in run
> 
>     and self.step()
> 
>         ^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 748, in step
> 
>     getattr(self, inst.opname)(inst)
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 467, in wrapper
> 
>     return inner_fn(self, inst)
> 
>            ^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 1825, in CALL
> 
>     self.call_function(fn, args, kwargs)
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/symbolic_convert.py", line 649, in call_function
> 
>     self.push(fn.call_function(self, args, kwargs))
> 
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/misc.py", line 878, in call_function
> 
>     return NumpyNdarrayVariable.create(tx, proxy)
> 
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/tensor.py", line 857, in create
> 
>     return wrap_fx_proxy_cls(
> 
>            ^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/variables/builder.py", line 1387, in wrap_fx_proxy_cls
> 
>     example_value = get_fake_value(proxy.node, tx, allow_non_graph_fake=True)
> 
>                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1585, in get_fake_value
> 
>     raise TorchRuntimeError(str(e)).with_traceback(e.__traceback__) from None
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1546, in get_fake_value
> 
>     ret_val = wrap_fake_exception(
> 
>               ^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1087, in wrap_fake_exception
> 
>     return fn()
> 
>            ^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1547, in <lambda>
> 
>     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
> 
>             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1652, in run_node
> 
>     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1631, in run_node
> 
>     return node.target(*args, **kwargs)
> 
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_dynamo/utils.py", line 1943, in __call__
> 
>     out = self.f(*args, **kwargs)
> 
>           ^^^^^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_numpy/random.py", line 46, in inner
> 
>     return func(*args, **kwds)
> 
>            ^^^^^^^^^^^^^^^^^^^
> 
>   File "/home/ken/Documents/GitHub/pytorch/torch/_numpy/random.py", line 111, in randn
> 
>     values = torch.randn(size, dtype=dtype)
> 
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> torch._dynamo.exc.TorchRuntimeError: Failed running call_function <Wrapped function <original randn>>(*(1000, 1000), **{}):
> 
> randn(): argument 'size' failed to unpack the object at pos 1 with error "AttributeError: 'int' object has no attribute 'is_int'"
> 
> ```
> 
> 
> 
> It's traced back to the lines
> 
> ```
> 
>             if isinstance(a, ConstantVariable) and proxy_for_constants and a.as_python_constant() is not None:
> 
>                 sym_a = torch.SymInt(a.as_python_constant())
> 
>                 const_proxy.node.meta["example_value"] = sym_a
> 
>                 sym_arg = SymNodeVariable(const_proxy, sym_a)
> 
>                 new_arg = sym_arg
> 
> ```
> 
> 
> 
> Seems that the numpy compatibility layer is treating a `SymInt` as a `SymNode`? Not sure why here.

Very interesting. Want me to TAL? Or are you going for it?

Fidget-Spinner commented:
I would greatly appreciate if you took a look. I took a stab and have barely any clue what is happening there. Thank you!

If you're busy I'll go for it instead.

voznesenskym commented:
> I would greatly appreciate if you took a look. I took a stab and have barely any clue what is happening there. Thank you!
> 
> 
> 
> If you're busy I'll go for it instead.

Off today for MLK, will look in the morning. I'll ping on here when I take the lock.

Once we solve this, let's fix the step thing, and then we can run the a100 performance jobs and see how this impacts the benchmarks suite.

Fidget-Spinner commented:
Thanks! Step seems to be fixed (I added a test for it)

voznesenskym commented:
https://github.com/pytorch/pytorch/actions/runs/7546095448

Fidget-Spinner commented:
Thanks for running the benchmark suite! We still need to fix the `SymNode` issue, else some numpy stuff might fail under dynamo.

Is there somewhere I can check out the collection of benchmarks? I'll be surprised if there are any benchmarks with loops large enough to benefit from this.

Fidget-Spinner commented:
Note: the closest commit this is to main is `afadfa0175221b0bb57af6f16fa2bb8210be7902`, which is on 29/28 Dec 2023 (depends on your timezone). I don't see any significant change in performance from the results.

Fidget-Spinner commented:
@voznesenskym after we fix the `SymNode` thing, I will decrease the config to size 1 and run the test suite again (this was suggested by Mario, thanks!). This will stress it more so we can find any latent bugs.

Fidget-Spinner committed:
symbolic int in range iterator

Fidget-Spinner committed:
[TEMP] stress by setting loop iter to 1

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner commented:
Test failures are strange. Code is not seeing `SymNodeVariable` as one in an `isinstance` check. Digging into this..

Fidget-Spinner committed:
block list/set/dictionary comprehensions

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner committed:
fix missing import

Fidget-Spinner commented:
Strange. I merge main in and the tests start failing. The graph produced is now different:

For the code:
```
        def f(x):
            res = x
            for i in range(loop_count):
                res = res + i + x
            return res
```

Before/expected:
```
    class GraphModule(torch.nn.Module):
        def forward(self, res, res_0, _lambda_):
            add = res_0 + _lambda_;  res_0 = None
            add_1 = add + res;  add = None
            return (res, add_1, _lambda_)
```
```
class GraphModule(torch.nn.Module):
    def forward(self, res, _lambda_):
    # File: /home/ken/Documents/GitHub/pytorch/test/dynamo/test_higher_order_ops.py:4041 in f, code: for i in range(loop_count):
        add = res + _lambda_
        add_1 = add + res;  add = None
        return (res, add_1, _lambda_)
```

Is there some sort of value numbering going on? The graph is correct in that `res0` and `res` are actually the same value. But eliminating it causes the forward to raise that number of arguments is wrong.

Fidget-Spinner committed:
remove accidental addition

lezcano commented:
@ezyang can you help finding someone that'll shepherd this one? Perhaps @mlazos? This one may need importing internally and all that.

ezyang commented:
If there aren't any volunteers, I guess @anijain2305 should get someone to do it.

I could review this in a pinch but I'm not too familiar with the implementation strategy here (so you'll have to page me in), and I am also somewhat unconvinced that, for speeding up loops, the first thing we should be doing is bytecode shenanigans, it doesn't necessarily seem like the right way to do it.

anijain2305 commented:
I can start getting up to speed on this in a couple of days. I would also ask @zou3519 and @jansel to review and also help with the direction.

Fidget-Spinner commented:
> I can start getting up to speed on this in a couple of days. I would also ask @zou3519 and @jansel to review and help with the direction.

This PR was working before I merged main in, then it broke the subgraph tracing subtlety. I will try to debug and fix this within the within one week.

zou3519 commented:
Yeah I can pick up

Fidget-Spinner commented:
Manually bisected the commit that broke this PR to https://github.com/pytorch/pytorch/commit/68105da229b4907207b55f50121389c8f570cc28

Looks legit. I will take a look soon.

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner committed:
(fix breakage) set subgraph inputs to manual

Fidget-Spinner committed:
block none,  don't lose precision

Fidget-Spinner committed:
fix tests, refactor

Fidget-Spinner committed:
lint

Fidget-Spinner committed:
deal with broken tests

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

zou3519 commented:
Btw, the internal diff looks OK. I'm going to try another one lowering the for-loop config count

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -421,9 +431,15 @@ def speculate_subgraph(
+                    if allow_constant_outputs
+                    else (TensorVariable, SymNodeVariable),
+                    allow_none=True,
+                ):
```
nit: update the unimplemented message depending on `allow_constant_outputs`

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
````diff
@@ -1196,6 +1213,329 @@ def call_function(
+    (i, x) = loop_body(1, x)
+    (i, x) = loop_body(2, x)
+    # ... 7 more times
+    ```
````
Do we handle the situation where a new local gets constructed in the body?

Fidget-Spinner replied:
In CPython, the new local gets "hoisted" out. So yes. What happens is before the loop it will be unset, but already "declared" (using these terms loosely as Python has no declarations).

zou3519 commented on torch/_dynamo/variables/lists.py:
```diff
@@ -701,6 +701,33 @@ class TupleIteratorVariable(ListIteratorVariable):
+
+    def next_variables(self, tx) -> "Tuple[VariableTracker, ListIteratorVariable]":
+        """
+        This should only be attempted once, so on the second time,
+        just return the original list iterator.
```
what do you mean?

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1166,8 +1171,50 @@ def POP_FINALLY(self, inst):
         it = self.pop().realize()
+        if (
+            isinstance(it, variables.RangeIteratorVariable)
+            and inst.starts_line is not None
```
what is `inst.starts_line is not None`?

zou3519 commented on torch/_dynamo/config.py:
```diff
@@ -374,6 +374,10 @@ def default_debug_dir_root():
+# How many iterations a for loop has to be before we decide to automatically
+# convert it to a function body.
+for_loop_medium_size_boundary = 100
+
```
should have a config option to disable this feature. Since it seems to test well we should start with a default of False in this PR and then flip it to True in a second PR.

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1166,8 +1171,50 @@ def POP_FINALLY(self, inst):
+                self.jump(inst)
+            except CannotConvertRangeToHigherOrder:
+                pass
+
```
I'm a bit confused at the flow. Let's say we succeed in creating the RangeHigherOrderVariable. Then we still go into the branch below (if isinstance(it, (variables.ListIteratorVariable, variables.IteratorVariable)):) because RangeIteratorVariable is a subclass of ListIteratorVariable. Is that intentional?

Fidget-Spinner replied:
Woops I forgot an `else` here. Thanks for catching this!

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1166,8 +1171,50 @@ def POP_FINALLY(self, inst):
+                        is RangeHigherOrderVariable.NOT_SET_SENTINEL
+                    ):
+                        continue
+                    self.symbolic_locals[name] = v
```
what happens if the body of the loop adds more locals on each iteration? Do we graph break successfully on that?

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1196,6 +1213,329 @@ def call_function(
+
+    # Any of these nested means the control flow is too complex to be
+    # represented by a higher order op.
+    FORBIDDEN_OPCODES = {
```
How difficult would it be to reverse this? That is, have a list of allowed opcodes instead of a list of forbidden opcodes.

The reason I ask because this doesn't seem very robust; if python adds a new opcode for something weird then it automatically gets allowed and we don't have good testing for it.

Fidget-Spinner replied:
I just tried and the list of opcodes is too long. We would need to list all opcodes from 3.8-3.10, and it changed once in 3.11, so that too, plus slight changes again in 3.12.

williamwen42 replied:
We can also abort early if we find an opcode in the loop body that can't be traced with symbolic_convert.

jansel replied:
Let's add this check as @williamwen42 suggested

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1196,6 +1213,329 @@ def call_function(
+        tx,
+        value: "RangeIteratorVariable",
+        real_globals: dict,
+        host_code_object: types.CodeType,
+        loop_body_instructions: List["Instruction"],
```
what is a host_code_object? The code object of the function the for loop is defined in?

Fidget-Spinner replied:
Yup!

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1196,6 +1213,329 @@ def call_function(
+            if sys.version_info >= (3, 11):
+                for _ in range(opcode._inline_cache_entries[inst.opcode]):
+                    co_code.extend((opcode.opmap["CACHE"], 0))
+        # We need to replace the last `JUMP_BACKWARD` with a RETURN_VALUE of all
+        # the locals.
```
are all FOR_ITER guaranteed to have a JUMP_BACKWARD?

Fidget-Spinner replied:
Yes, if i understand things correctly.

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1196,6 +1213,329 @@ def call_function(
+        obj.store_target = store_target
+        return obj
+
+    def functionalize(self, tx) -> tuple:
```
nit: funcitonalize means something specific in AOTAutograd (the process of turning in-place operations into out-of-place operatiosn). This seems more like a "we're gonna put the for-loop into the graph under our custom scheme"

zou3519 reviewed (commented):
I need to run, but the general approach seems reasonable to me. We should get more eyes on what's going on in working with the code object and banning opcodes.

I wonder if there's a better way we can test this.

Fidget-Spinner committed:
add another test

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner committed:
fix upstream merges

Fidget-Spinner commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1196,6 +1213,329 @@ def call_function(
+
+        if code_object is None:
+            raise CannotConvertRangeToHigherOrder("Could not construct code object")
+
```
Yup!

Fidget-Spinner committed:
address feedback

anijain2305 commented on torch/_dynamo/config.py:
```diff
@@ -408,6 +408,10 @@ def default_debug_dir_root():
+# How many iterations a for loop has to be before we decide to automatically
+# convert it to a function body.
+for_loop_medium_size_boundary = 100
+
```
Please add a config option to disable this feature ... we can gradually turn it on.

anijain2305 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+    better than what we currently have.
+    """
+
+    # Any of these nested means the control flow is too complex to be
```
cc @williamwen42 for the list of fbcodes if he sees anything is missing.

anijain2305 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        varnames = host_code_object.co_varnames
+        args = [
+            symbolic_locals.get(k)
+            or ConstantVariable.create(RangeHigherOrderVariable.NOT_SET_SENTINEL)
```
What is this for?

Fidget-Spinner replied:
Yup William is right. a variable might not yet be defined.

anijain2305 reviewed (changes requested):
Thanks for the great work! Looks good to me.

Please do use a config flag to control this. For this PR, turn this flag to off (and for tests, mock it to True). And then a second PR where you turn the flag on. Just easy to deal with reverts.

williamwen42 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        # STORE_FAST always follows a FOR_ITER as CPython needs to store the next(iter) into the local
+        # E.g. in `for i in range(10)`, there is a `STORE_FAST i``
+        assert loop_body_instructions[0].opname == "STORE_FAST"
+        co_code: List[int] = []
```
Any particular reason why we're avoiding `torch._dynamo.bytecode_transformation` in order to create the new code object? It provides some nice utilities such as dealing with CACHE entries and utility functions to create instructions. 

e.g. https://github.com/pytorch/pytorch/blob/4ceb44c40d52c57143ab1dbc8c14373f5205e9fb/torch/_dynamo/resume_execution.py#L492

Fidget-Spinner replied:
Didn't know of that, will use it !

williamwen42 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        co_code.extend(
+            (
+                opcode.opmap["BUILD_TUPLE"],
+                host_code_object.co_nlocals,
```
What if we need to use an extended arg here because we have a lot of locals? Using `bytecode_transformation` should deal with this case properly.

williamwen42 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        "YIELD_FROM",
+        "GET_YIELD_FROM_ITER",
+        "SEND",
+        "MAKE_FUNCTION",
```
Can we support making functions that are not closures?

Fidget-Spinner replied:
I think we can. But I'll leave that for future work.

williamwen42 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+    # Any of these nested means the control flow is too complex to be
+    # represented by a higher order op.
+    FORBIDDEN_OPCODES = {
+        "RETURN_VALUE",
```
RETURN_CONST is needed for 3.12.

williamwen42 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        # These two are fine for the end of the loop body but not within.
+        # TODO: we can support these too. They just need to be transformed to the
+        # right RETURN_VALUEs (see code below for how to do that).
+        "JUMP_BACKWARD",
```
Do we need to add other jump opcodes like `JUMP_FORWARD`, `POP_JUMP_IF_FALSE`, etc.

Fidget-Spinner replied:
No. `speculate_subgraph` handles stuff like if-else IIUC. That should be guarded.

williamwen42 replied:
I think `JUMP_FORWARD` and `POP_JUMP_IF_FALSE` can escape the scope of the for loop, e.g. if we break - is that handled?

Fidget-Spinner replied:
Yes the `POP_JUMP_IF_FALSE` previously jumped to `JUMP_BACKWARDS`, now it jumps to `BUILD_TUPLE + RETURN_VALUE`.

williamwen42 replied:
I still think we need to add more jump opcodes, e.g. in 3.10:
```python
>>> def fn(x):
...     for i in range(10):
...             x += i
...             if i > 5:
...                     x += i
...                     break
...             x += 1
...     x += 1
...     return x
... 
>>> dis.dis(fn)
  2           0 LOAD_GLOBAL              0 (range)
              2 LOAD_CONST               1 (10)
              4 CALL_FUNCTION            1
              6 GET_ITER
        >>    8 FOR_ITER                20 (to 50)
             10 STORE_FAST               1 (i)

  3          12 LOAD_FAST                0 (x)
             14 LOAD_FAST                1 (i)
             16 INPLACE_ADD
             18 STORE_FAST               0 (x)

  4          20 LOAD_FAST                1 (i)
             22 LOAD_CONST               2 (5)
             24 COMPARE_OP               4 (>)
             26 POP_JUMP_IF_FALSE       20 (to 40)

  5          28 LOAD_FAST                0 (x)
             30 LOAD_FAST                1 (i)
             32 INPLACE_ADD
             34 STORE_FAST               0 (x)

  6          36 POP_TOP
             38 JUMP_FORWARD             5 (to 50)

  7     >>   40 LOAD_FAST                0 (x)
             42 LOAD_CONST               3 (1)
             44 INPLACE_ADD
             46 STORE_FAST               0 (x)
             48 JUMP_ABSOLUTE            4 (to 8)

  8     >>   50 LOAD_FAST                0 (x)
             52 LOAD_CONST               3 (1)
             54 INPLACE_ADD
             56 STORE_FAST               0 (x)

  9          58 LOAD_FAST                0 (x)
             60 RETURN_VALUE
```
we are `JUMP_FORWARD`ing past the `JUMP_ABSOLUTE`.

williamwen42 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        # This could be supported in the future.
+        "LOAD_DEREF",
+        # Exceptions
+        "RAISE_VARARGS",
```
Exceptions in 3.11+ are trickier to deal with because bytecodes denoting the start and end of a try block are gone and instead are replaced by an exception table.

3.12 in particular is really nasty because the except block bytecode is placed after the END_FOR.

Summary: we need to pay attention to the exception table for 3.11+.

williamwen42 replied:
Are you also forbidding with blocks too? Because they seem to contain `POP_EXCEPT` in 3.9+. But in 3.8, with blocks don't contain any of `RAISE_VARARGS, SETUP_FINALLY, POP_EXCEPT`.

I still think in 3.11+, you will need to look at the exception table. If with blocks are forbidden, then you just need forbid instructions that have an exception table entry. If with blocks are not forbidden, then you will have to additionally determine if the exception table entry corresponds to a with block or not, by looking for the `BEFORE_WITH` instruction.

Fidget-Spinner replied:
I'm not but I think we can start simple for now. I've banned anything containing an exception table entry for now.

Fidget-Spinner commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        code_object = None
+        # Translate instructions back to code object.
+        if sys.version_info >= (3, 12):
+            raise CannotConvertRangeToHigherOrder("Unsupported Python version")
```
Lintrunner is failing I think. I can support 3.12 now, just need to play with the codetype.

Fidget-Spinner commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        # right RETURN_VALUEs (see code below for how to do that).
+        "JUMP_BACKWARD",
+        "JUMP_ABSOLUTE",
+        # list, set, dictionary comprehensions
```
A comprehension is itself a loop. Nested loops are not supported for now, but can be in the future.

williamwen42 replied:
I was worried these instructions could be used elsewhere but it looks like these instructions exist specifically for comprehensions.

Fidget-Spinner commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1217,8 +1222,56 @@ def POP_FINALLY(self, inst):
+                # Also pop off the iterator.
+                self.jump(inst)
+                # leave iterator upon exhaustion in 3.12
+                if sys.version_info >= (3, 12):
```
otherwise it's not executed in 3.12, since the code above skips the entire block below

williamwen42 replied:
Can we refactor into 1 function?

Also, I don't see why the try block below
```python
        try:
            val = it.next_variable(self)
            self.push(it)
            self.push(val)
```
is skipped - it looks like in 3.12, we'd be pushing 4 things.

Fidget-Spinner commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,329 @@ def call_function(
+        assert loop_body_instructions[0].opname == "STORE_FAST"
+        co_code: List[int] = []
+        if sys.version_info >= (3, 11):
+            co_code.extend((opcode.opmap["RESUME"], 0))
```
Yep that's blocked by the forbidden opcode list. That checks for `YIELD_VALUE`.

Fidget-Spinner committed:
Address William's and Animesh's reviews

Fidget-Spinner committed:
lint

Fidget-Spinner committed:
fix failing test

Fidget-Spinner committed:
refactor as requested

Fidget-Spinner commented:
@anijain2305 and @jansel gentle ping for a review please. Thank you!

williamwen42 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,286 @@ def call_function(
+        # the locals.
+        for i in range(host_code_object.co_nlocals):
+            co_code.append(
+                Instruction(opcode.opmap["LOAD_FAST"], "LOAD_FAST", i, varnames[i])
```
You can try using `bytecode_transformation.create_instruction` like so: `create_instruction("LOAD_FAST", argval=varnames[i])`. Any required instruction args should be automatically converted from argvals by `bytecode_transformation`.

Fidget-Spinner committed:
ban exception table entries, use create_instruction

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

jansel commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,286 @@ def call_function(
+        "SEND",
+        "MAKE_FUNCTION",
+        # This could be supported in the future.
+        "LOAD_DEREF",
```
What about STORE_DEREF/DELETE_DEREF/etc.

Fidget-Spinner replied:
I think those will be handled by `speculate_subgraph`, but just to be sure I will add them. In theory a lot of things can be handled by `speculate_subgraph`. For example, it makes code in a `try-except` safe for this transformation too. But just to be safe, I will block anything that touches exceptions or closures.

jansel commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,286 @@ def call_function(
+        "MAP_ADD",
+        # 3,12
+        "RETURN_CONST",
+    }
```
I'm a bit worried we missed something here...

jansel commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1099,6 +1117,286 @@ def call_function(
+        assert loop_body_instructions[-1].opname in {"JUMP_BACKWARD", "JUMP_ABSOLUTE"}
+        if loop_body_instructions[0].opname == "UNPACK_SEQUENCE":
+            raise CannotConvertRangeToHigherOrder(
+                "Unpacking a range iterator, seems wrong."
```
Why is this wrong?  Isn't it just:
```
for a,b in ...
```
which is common

Fidget-Spinner replied:
This conversion only currently works for `range(x)`. There's no way to unpack a range iterator like that.

jansel reviewed (changes requested):
Please do a full CI run with this enabled and the threshold set to 1.  I worry there are correctness issues.

Fidget-Spinner commented:
> Please do a full CI run with this enabled and the threshold set to 1. I worry there are correctness issues.

I already did that earlier. It fails CI because we have tests asserting that the number of graphs is a certain count, which changes with this PR. Apart from those failures, there aren't correctness failures that I am aware of.

@zou3519 imported this to the internal codebase and ran the internal test suite and apparently it passes, (bar the usual same failures mentioned above).

Fidget-Spinner commented on torch/_dynamo/bytecode_transformation.py:
```diff
@@ -1147,9 +1147,6 @@ def clean_and_assemble_instructions(
 
     code_options["co_code"] = bytecode
     code_options["co_stacksize"] = stacksize_analysis(instructions)
-    assert set(keys) - {"co_posonlyargcount"} == set(code_options.keys()) - {
```
Good question. We now create code objects that do not inherit all the flags of the parent object. For example, in this case I'm creating a for loop function that is just a simple function, but it could be inside a generator function. In this case I don't want to inherit the flags of the generator function for the for loop body.

Fidget-Spinner committed:
Address review by Jason

Fidget-Spinner commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1217,9 +1222,48 @@ def POP_FINALLY(self, inst):
+                    new_locals = op.to_function(self)
+                    for loc in new_locals:
+                        if isinstance(loc, torch.fx.Proxy):
+                            del loc.node.meta["example_value"]
```
I'm not sure. Will delete this.

Fidget-Spinner committed:
avoid mutation

Fidget-Spinner committed:
lint

Fidget-Spinner commented:
I think all the reviews have been addressed other than Jason's concern we might be missing some bytecode, which I'm not sure how to address other than looking at the dis module again later one more time.

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner commented:
@jansel and @anijain2305 gentle ping please. I think I addressed all reviews that I am able to address. I'm not able to run this on CI fully enabled and expect it to pass because we have tests things like frame count and operation count which this obviously changes.

lezcano commented:
There are some errors that seem unrelated. See [[this one](https://github.com/pytorch/pytorch/actions/runs/9298295797/job/25591369444)]:
>2024-05-30T09:30:07.9956022Z torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in method conv2d of type object at 0x7fcf75a1d660>(*(FakeTensor(..., size=(2, 3, 5, 5)), FakeTensor(..., size=(3, 3, 1, 1), grad_fn=<Invalid>), FakeTensor(..., size=(3,), grad_fn=<Invalid>), (1, 1), (0, 0), (1, 1), 1), **{}):
2024-05-30T09:30:07.9958453Z Output 0 of TracableCreateParameterBackward is a view and its base or another view of its base has been modified inplace. This view was created inside a custom Function (or because an input was returned as-is) and the autograd logic to handle view+inplace would override the custom backward associated with the custom Function, leading to incorrect gradients. This behavior is forbidden. You can fix this by cloning the output of the custom Function.
2024-05-30T09:30:07.9959136Z Exception raised from handle_view_on_rebase at /var/lib/jenkins/workspace/torch/csrc/autograd/variable.cpp:852 (most recent call first):

Can you add the relevant changes so that everything passes to make sure we are not missing anything? You can keep a local diff to revert everything before landing (with ghstack all this would be easier, but well)

Fidget-Spinner committed:
update tests

jansel commented:
I'd like to see some more benchmarking numbers here on real models and with different thresholds.  Does this make a meaningful improvement in compile time?  Does this affect performance?  Running on real models with a small threshold (like 3) might also shake out some bugs.

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner commented:
Voz once ran a benchmark on the Torchbench suite and it reported no speedups. I attribute it to two reasons:

1. Not many large for loops in the benchmarks.
2. For loops that are too complex to represent for now. This can be fixed in the future if a for loop HOP that support side effects is introduced.

To see how much this improved code, I went to this issue reporting slow for loop compiles https://github.com/pytorch/pytorch/issues/115626. The user suggests this is a slimmed down version of a real model they are using.

I then took the code, modified it a little so it ran in a reasonable time on my computer, and benchmarked it. Results are at this gist https://gist.github.com/Fidget-Spinner/f592d900e32cc05f5cabf70f3b519c21.

Total program time went from 3m17.354s on main to 2m36.041s on my branch. Representing a ~20% speedup. This implies for loops sped up by more than 20% as that number is total program execution time, not just compilation time.

Fidget-Spinner commented:
@jansel are there specifically any other loop benchmarks you would like to see? The one I gave above was apparently a shortened version from a user's actual code. And it showed a ~20% speedup.

jansel commented:
I was looking for more full model benchmarks.

Fidget-Spinner commented:
> I was looking for more full model benchmarks.

Well sorry about that but Torchbench shows no speedups. I explained more here https://github.com/pytorch/pytorch/pull/113538#issuecomment-2141198686

jansel commented:
Are there any real models where this shows speedups?

Fidget-Spinner commented:
Not that I'm aware of. But considering there are two open issues on slow for loop compilation I still think it's worth it.

lezcano commented:
And also note that the idea is to have this just trigger on very long loops, and just change it to trigger on small loops once we have function caching at a dynamo level

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/main' into recursive_definitions

Fidget-Spinner commented:
@jansel 

Would you like to make a final decision in this PR? (Accept/Reject). I'm fine either ways. After all the outcome should be in line with what the project needs.

Just laying out the pros and cons again of this PR for your consideration:

Pros:
* >= 20% faster loop compilation
* Lays the foundation for automatic O(1) loop compilation

Cons:
* Requires some fiddling with CPython bytecode.
* More lines of code and the associated maintenance burden
* Does not support all loops (yet, this is a temporary con and can be fixed in the future)

jansel reviewed (changes requested):
I'm leaning no on this one.  This PR adds a lot of code complexity and I don't see the compile time gains manifesting in real models.

For hierarchical compilation, I'm not sure this approach is needed.  Unusually code is structured with a repeated layer object that is already a separate function call.

lezcano commented:
Just for reference, loops are not common on deep learning models, but they are quite common on scientific simulations, like the example linked in the OP

Fidget-Spinner commented:
While I would like to see this PR merged, I respect Jason's decision here. So I'm closing this.
